enhancements/operator-conditions-lib.md

```diff
@@ -0,0 +1,144 @@
+
+2. Specify conditions in the CRD:
+
+The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom ([non-OLM supported][non-olm-supported-condition]) condition for communicating the status of the operator to users. An example Status CR is given below:
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: Status
+metadata:
+  name: foo-operator-zdgen-asdf # RandomGen Name
+  namespace: operators
+status:
+  conditions:
+    type: Upgradeable # The name of the `foo` OLM Supported Condition.
+    status: "False"   # The operator is not ready to be upgraded.
+    reason: "migration"
+    message: "The operator is performing a migration."
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the [`status`][status-sub-resource] sub-resource of the CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+### Scenarios:
+
+1. Operator does not set/update any conditions that affect OLM:
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update the `Condition` CR when it is not present in the cluster:
+
+In this case, the specific API will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+3. Operator tries to get/update condition CRD which is present in cluster, but the requested CR instance is not available.
+
+In this case, the error returned will be in the same format as stated in the previous scenario but would specifically mention that the requested object is not found.
+
+To summarize, these are the overall helper methods/functions present in the library. For the ease of understanding, in the below table:
+- `Status` refers to the CR definition provided by OLM and 
+- [`metav1.Condition`][metav1-Condition] refers to specific Condition type present in `status.conditions` array.
+
+ Description             | Input(s)      | Return type(s) |  Expected output  | Remarks    |
+-------------            | ------------- | -------------| ----------------  |------------- |
+  GetStatusCRD        | Context, Runtime Object (`Status`), Object Key | `Status` struct, error | `Status` provided in input populated with values obtained from cluster |  1. Library will have helper functions to construct Object key <br/>2. Return errors as discussed in Scenarios section |
```

Comment:
Can we just provide a library func that gets the `types.NamespacedName` for the `Condition` CRD instead of a `Get()` wrapper? `GetStatusCRD()` would have to either be a method on a type wrapping a `Client` or have a `Client`-type parameter, which are ugly imo.

Reply:
Does this mean, the library will not provide a method to get the entire CRD from cluster? Instead we will only provide a helper to get `types.NamspacedName` and the operator would use `Get` client to fetch the CRD from cluster?

Reply:
Correct:
```go
// Get key.Name from env var, key.Namespace from downward API.
key, err := libcondition.GetConditionNamespacedName()
if err != nil {
	return err
}

condition := v1alpha1.Condition{}
if err = r.client.Get(ctx, key, &condition); err != nil	{
	return err
}
...
```

Reply:
Reworded the EP a bit and included this.